src/Dependencies/PooledObjects/ArrayBuilder.cs

```diff
@@ -470,6 +470,16 @@ public void AddRange(T[] items, int length)
             _builder.AddRange(items, length);
         }
 
+        public void AddRange<TSource>(TSource[] items, Func<TSource, T> selector)
```

Comment:
➡️ Avoids an intermediate array allocation, and ensures the resizing of the target will occur at most once for this operation.